File: ai agent/material/python/11_class_inheritance.py
class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        return "Some sound"

    def move(self):
        return f"{self.name} is moving"

class Dog(Animal):
    def speak(self):
        return "Woof!"

    def fetch(self):
        return f"{self.name} is fetching the ball"

class Cat(Animal):
    def speak(self):
        return "Meow!"

    def climb(self):
        return f"{self.name} is climbing a tree"

class Bird(Animal):
    def __init__(self, name, can_fly=True):
        super().__init__(name)
        self.can_fly = can_fly

    def speak(self):
        return "Chirp!"

    def move(self):
        if self.can_fly:
            return f"{self.name} is flying"
        return f"{self.name} is walking"

animals = [Dog("Buddy"), Cat("Kitty"), Bird("Tweety")]
for a in animals:
    print(f"{a.name}: {a.speak()} - {a.move()}")

File: projects/ai-agent/material/python/cpp/11_class_inheritance.cpp
#include <iostream>
#include <string>
using namespace std;

class Animal {
protected:
    string name;

public:
    Animal(string n) : name(n) {}
    virtual void speak() {
        cout << name << " makes a sound" << endl;
    }
    virtual ~Animal() {}
};

class Dog : public Animal {
public:
    Dog(string n) : Animal(n) {}
    void speak() override {
        cout << name << " says: Woof!" << endl;
    }
    void fetch() {
        cout << name << " fetches the ball" << endl;
    }
};

class Cat : public Animal {
public:
    Cat(string n) : Animal(n) {}
    void speak() override {
        cout << name << " says: Meow!" << endl;
    }
};

int main() {
    Animal* animals[3];
    animals[0] = new Dog("Buddy");
    animals[1] = new Cat("Kitty");
    animals[2] = new Dog("Max");

    for (int i = 0; i < 3; i++) {
        animals[i]->speak();
    }

    for (int i = 0; i < 3; i++) {
        delete animals[i];
    }

    return 0;
}
